health-events-analyzer/pkg/parser/parser.go

```diff
@@ -0,0 +1,217 @@
+	doc := bson.D{}
+
+	for key, value := range criteria {
+		// Non-string values are appended as-is
+		strVal, isString := value.(string)
+
+		if !isString {
+			doc = append(doc, bson.E{Key: key, Value: value})
+			continue
+		}
+
+		// "this." reference → resolve from current event
+		if strings.HasPrefix(strVal, "this.") {
+			fieldPath := strings.TrimPrefix(strVal, "this.")
+			resolvedValue, err := p.getValueFromPath(fieldPath)
+
+			if err != nil {
+				return doc, fmt.Errorf("error in getting value from path: %w", err)
+			}
+
+			doc = append(doc, bson.E{Key: key, Value: resolvedValue})
+
+			continue
+		}
+
+		// JSON object string representing MongoDB operator (e.g. '{"$ne":"x"}')
+		if strings.HasPrefix(strVal, "{") && strings.HasSuffix(strVal, "}") {
+			var operatorMap map[string]any
+			if err := json.Unmarshal([]byte(strVal), &operatorMap); err != nil {
+				slog.Warn("Failed to parse MongoDB operator string, treating as literal", "string", strVal, "error", err)
+				doc = append(doc, bson.E{Key: key, Value: strVal})
+			} else {
+				doc = append(doc, bson.E{Key: key, Value: operatorMap})
+			}
+
+			continue
+		}
+
+		// Plain string literal
+		doc = append(doc, bson.E{Key: key, Value: strVal})
+	}
+
+	return doc, nil
+}
+
+// getValueFromPath extracts a value from the event using a dot-notation path
+func (p *Parser) getValueFromPath(path string) (any, error) {
+	parts := strings.Split(path, ".")
+
+	if len(parts) > 0 && (parts[0] == "healthevent") {
+		return getValueFromHealthEvent(p.Event.HealthEvent, parts[1:]), nil
+	} else if len(parts) > 0 && (parts[0] == "healtheventstatus") {
+		return getValueFromHealthEventStatus(p.Event.HealthEventStatus, parts[1:]), nil
+	}
+
+	return nil, fmt.Errorf("invalid path: %s", path)
+}
+
+// getFieldByName is a common helper function to find a field by name using reflection (case-insensitive)
+func getFieldByName(val reflect.Value, fieldName string) any {
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Type().Field(i)
+		if strings.EqualFold(field.Name, fieldName) {
+			return val.Field(i).Interface()
+		}
+	}
+
+	return nil
+}
+
+func getValueFromHealthEvent(event *platform_connectors.HealthEvent, parts []string) any {
+	if len(parts) == 0 {
+		return nil
+	}
+
+	rootField := strings.ToLower(parts[0])
+
+	// Simple one-level field lookup
+	if len(parts) == 1 {
+		return getFieldByName(reflect.ValueOf(event).Elem(), rootField)
+	}
+
+	switch rootField {
+	case "errorcode":
+		return getValueFromErrorCode(event, parts[1:])
+	case "entitiesimpacted":
+		return getValueFromEntitiesImpacted(event, parts[1:])
+	case "metadata":
+		return getValueFromMetadata(event, parts[1:])
+	case "generatedtimestamp":
+		return getValueFromGeneratedTimestamp(event, parts[1:])
+	default:
+		return nil
+	}
+}
+
+// getValueFromErrorCode safely returns event.ErrorCode[index] if present.
+func getValueFromErrorCode(event *platform_connectors.HealthEvent, parts []string) any {
```

Comment:
should we rename all the function: `getValueFromX` -> `getX`? Will it more readable?